cc/mcts_player.h

```diff
@@ -0,0 +1,145 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CC_MCTS_PLAYER_H_
+#define CC_MCTS_PLAYER_H_
+
+#include <cmath>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include "absl/memory/memory.h"
+#include "absl/types/span.h"
+#include "cc/algorithm.h"
+#include "cc/constants.h"
+#include "cc/dual_net.h"
+#include "cc/mcts_node.h"
+#include "cc/position.h"
+#include "cc/random.h"
+
+namespace minigo {
+
+class MctsPlayer {
+ public:
+  struct Options {
+    bool inject_noise = true;
+    bool soft_pick = true;
+    bool random_symmetry = true;
+
+    float resign_threshold = -0.9;
+
+    int batch_size = 8;
+
+    float komi = kDefaultKomi;
+
+    // Seed used from random permutations.
+    // If the default value of 0 is used, a time-based seed is used.
+    uint64_t random_seed = 0;
+  };
+
+  struct History {
```

Comment:
thoughts on whether you want to directly output tfrecords or whether you want to output the search_pi array and postprocess them into tf.records via existing python code? IMO, latter sounds a bit less efficient but also less error prone. It's difficult to detect errors in the data transcription process; we've previously only noticed errors after our training started going wonky after a few days.

Reply:
I hadn't given it much thought yet to be honest. I share your concern about potential bugs; one option would be to use the Python code to generate a set of golden outputs to compare the C++ code to.